contracts/poap/src/integration_tests.rs

```diff
@@ -0,0 +1,349 @@
+                height: 42,
+                time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+                chain_id: "testchain".to_string(),
+            })
+            .build(|_, _, _| {})
+    }
+
+    /// Uploads the contracts to the app.
+    /// Returns a pair of ids where the first refers to the cw721
+    /// and the second to the poap.
+    fn store_contracts(app: &mut App) -> (u64, u64) {
+        let cw721_code_id = app.store_code(cw721_test_utils::contract_cw721());
+        let poap_code_id = app.store_code(contract_poap());
+
+        (cw721_code_id, poap_code_id)
+    }
+
+    fn proper_instantiate() -> (App, Addr) {
+        let mut app = mock_app();
+        let (cw721_code_id, poap_code_id) = store_contracts(&mut app);
+        let msg = get_valid_init_msg(cw721_code_id);
+
+        let poap_contract_addr = app
+            .instantiate_contract(
+                poap_code_id,
+                Addr::unchecked(ADMIN),
+                &msg,
+                &[],
+                "Poap contract",
+                None,
+            )
+            .unwrap();
+
+        (app, poap_contract_addr)
+    }
+
+    #[test]
+    fn instantiate_with_invalid_cw721_code_id() {
+        let mut app = mock_app();
+        let (cw721_code_id, poap_code_id) = store_contracts(&mut app);
+        let mut init_msg = get_valid_init_msg(cw721_code_id);
+
+        init_msg.cw721_code_id = 42u64.into();
+
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+    }
+
+    #[test]
+    fn instantiate_with_failing_cw721_contract() {
+        let mut app = mock_app();
+        let (cw721_code_id, poap_code_id) = store_contracts(&mut app);
+        let failing_cw721_code_id = app.store_code(cw721_test_utils::failing_cw721());
+        let mut init_msg = get_valid_init_msg(cw721_code_id);
+
+        init_msg.cw721_code_id = failing_cw721_code_id.into();
+
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+    }
+
+    #[test]
+    fn proper_contracts_instantiation() {
+        let (app, poap_contract_addr) = proper_instantiate();
+
+        let querier = app.wrap();
+
+        let poap_config: QueryConfigResponse = querier
+            .query_wasm_smart(&poap_contract_addr, &QueryMsg::Config {})
+            .unwrap();
+
+        assert_eq!(Addr::unchecked(ADMIN), poap_config.admin);
+        assert_eq!(Addr::unchecked(MINTER), poap_config.minter);
+        assert_eq!(false, poap_config.mint_enabled);
+        // 1 since is the first uploaded.
+        assert_eq!(Uint64::new(1), poap_config.cw721_contract_code);
```

Comment:
I think it should be tested in contract test. The integration test should be focus on the states of other contracts.

Reply:
This tests the cw721 state after initialization and ensure that also the poap contract have a valid state after initizalization, also this contract in order to be properly initialized needs to receive a `MsgInstantiateContractResponse` as a replay data, so imo this is an integration test.

If we want to convert this into a unit test we need to manually call `replay` and serialize `MsgInstantiateContractResponse` but looking into cw-utils I can't find any utility function to properly serialize it, we can obviosly write our own proto serialization but seems just a dirt hack imo, if in the future they change this struct all our tests will probably start failing and we also need to manage some code that do the serialization of a external library struct.

Reply:
Yeah, you are right, it is needed a `MsgInstantiateContractResponse`. I covered these reply test in the POAP manager contract with [structure](https://github.com/desmos-labs/desmos-contracts/blob/paul/DCD-51/poap-manager-impl/contracts/poap-manager/src/contract.rs#L238) and reply test[ here](https://github.com/desmos-labs/desmos-contracts/blob/paul/DCD-51/poap-manager-impl/contracts/poap-manager/src/contract.rs#L390). What do you think for this way?

Reply:
I see, as I said this aproach can lead to problems in the future if the `MsgInstantiateContractResponse` changes.
We have the integration tests framework that takes care of this for apart for a strong motivation i don't see why don't use that.

Reply:
I think it's okay to test the reply working properly (if not already tested by `cosmwasm` itself) but I think it's useful to have also this integration test since we are calling different components and not testing units only. I would leave this test here if possible.

Reply:
I see, let's remain the reply msg check in integration test.